SHARE-154 Detail page for Scratch projects

11possim opened the pull request:
Added the corresponding scratch IDs to program and user entities.
Added /scratch_project/{id} endpoint, it creates a catrobat project (if there isn't one already) and redirects to that project.
A user is also added (if it doesn't already exist). The usernames for Scratch users are: 'Scratch:<Name>'.
Added a 'Profile on Scratch' Button in profile view of scratch users.
Didn't change program frontend b/c currently in redesign (Share-267).
Remix relations aren't added at the moment.
Added the /scratch_project/{id} endpoint to robots.txt to prevent the creation of new projects by crawler.

---
### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroweb-Symfony/blob/develop/.github/contributing.md) and [wiki pages](https://github.com/Catrobat/catroweb-Symfony/wiki/) of this repository.

- [x] Include the name and id of the Jira ticket in the PR’s title eg.: `SHARE-666 The devils ticket`
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no warnings and errors 
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Verify that all tests are passing, if not please state the test cases in the [section](#Tests) below
- [x] Perform a self-review of the changes
- [x] Stick to the project’s git workflow (rebase and squash your commits)
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] Put your ticket into the `Code Review` section in [Jira](https://jira.catrob.at/)
- [ ] Post a message in the *#catroweb* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### Additional Description
`TODO: Add additional information that is not in your commit-message here`

### Tests - additional information
`TODO: add additional information about testruns here`

11possim committed:
SHARE-154 Detail page for Scratch projects

Added the corresponding scratch IDs to program and user entities.
Added /scratch/project/{id} endpoint, it creates a catrobat project (if there isn't one already) and redirects to that project.
A user is also added (if it doesn't already exist). The usernames for Scratch users are: 'Scratch:<Name>'.
Added a 'Profile on Scratch' Button in profile view of scratch users.
Didn't change program frontend b/c currently in redesign (Share-267).
Remix relations aren't added at the moment.
Added the /scratch endpoint to robots.txt to prevent the creation of new projects, user etc. by crawler.

dmetzner commented:
It works like a charm! ~Even the download, which was not required by the ticket, works.~

- ~**missing indication** that it's a scratch project other than the username. 
` The detail page should include a static, translated text right after the description, that mentions that this is a Scratch project, and includes the link to it.` I guess also because of  (Share-267)?
At least we should **create a new ticket** for it. Else we will forget about it.~

- The **Scratch User profile button** ~looks a bit weird. I'll contact UX and ask them before I say something wrong:~ Please add some spacing in between for the mobile view. The button will find it's final position in the profile redesign. (UX updates the mockups)

![image](https://user-images.githubusercontent.com/40868718/85100133-a1599180-b1ff-11ea-9d88-607855d74a9c.png)

## Some thoughts for discussion:
@11possim @bernadettespieler @HCrane 

- `/scratch_project` route - should we maybe change it to `/scratch/project` to be more compatible with possible future additions? like `/scratch/user` to import whole users from scratch 

- the logic behind `/scratch_project` must be added to the CatrowebAPI. **create a new ticket?**

- ~should we prevent username creation with a colon sign? So only this autogenerated usernames can have the `Scratch:` prefix. Else I could create an account 'Scratch:ABC" and "steal" ABC identity/projects.~ **Update Capi Documentation**

- `From the ticket description: "The links to the detail pages of Scratch projects on our server should NOT be shown in any section of the main page, i.e. not in the “Newest projects” section etc."`
   Why would we want this? After the conversion, the project is basically is a Catrobat project. If it is loved by many people or a cool game, why shouldn't we recommend it to others? Plus isn't the scratch category on the landing page explicitly for such projects?

- ~What happens if the scratch user updates his project on scratch? Can we somehow check out the new updated version of the project~ The updated question - when and how can this conversion be done? Users will never visit the  `/scratch_project` URL. How can they convert projects?  (happens this only during the search?")

- ~should we also "steal" the likes/reactions, views, downloads, comments from scratch projects, and add them to our catrobat version project?~

dmetzner commented on src/Api/UserApi.php:
```diff
@@ -124,6 +124,10 @@ public function validate(Register $register): ValidationSchema
     {
       $validation_schema->setUsername($this->translator->trans('api.registerUser.usernameAlreadyInUse', [], 'catroweb'));
     }
+    elseif (0 === strncmp($register->getUsername(), User::$SCRATCH_PREFIX, strlen(User::$SCRATCH_PREFIX)))
```
AH nice, you already thought about this! :)
So, we just need to update the CAPI documentation

dmetzner commented on src/Entity/ProgramManager.php:
```diff
@@ -307,6 +300,80 @@ public function addProgram(AddProgramRequest $request): ?Program
+      }
+      $program->incrementVersion();
+    }
+    $program->setCatrobatVersion(1);
```
I'm not sure about this version number. Hardcoding it will result in many problems. Catblocks + IOS team depend on this number.

11possim replied:
All usages of setCatrobatVersion have 1 as parameter. Maybe we should delete that column?
The info about the catrobat version seems to be in CatrobatVersionName.
Should that one be set even if no project file currently exists?

dmetzner replied:
Yeah, for sure, let's delete it. But it can be done on a different ticket. It would be necessary to rework this version feature anyway. Instead of the string should use the typical major.minor.patch format. The sql string comparison right now is a little bit hacky - requiring exactly the correct format (same string length) to work. **Ticket needed**

dmetzner replied:
No if there is no project it should not be set.

11possim replied:
Added [SHARE-333](https://jira.catrob.at/browse/SHARE-333)

dmetzner commented on src/Entity/ProgramManager.php:
```diff
@@ -307,6 +300,80 @@ public function addProgram(AddProgramRequest $request): ?Program
+    $program->setCatrobatVersion(1);
+    $program->setVisible(true);
+    $program->setApproved(false);
+    $program->setCatrobatVersion(1);
```
+ why set it twice ;)

dmetzner commented on src/Catrobat/Services/OAuthService.php:
```diff
@@ -343,7 +343,7 @@ public function loginWithTokenAndRedirectAction(Request $request): JsonResponse
   }
 
-  public function randomPassword(): string
+  public static function randomPassword(): string
```
we should move this to some utils

dmetzner commented on tests/behat/features/web/project-details-3/project_scratch.feature:
```diff
@@ -0,0 +1,32 @@
+    And there should be "1" users in the database
+
+  Scenario: Visiting a not existing Scratch program
+    #Given I request "GET" "/app/scratch_project/1"
```
unused code (comment) which can be removed

dmetzner reviewed (changes requested):
You also have changed almost all shell scripts. Any reason for this "0 lines, but bytes have changed" changes?

11possim commented:
> Even the download, which was not required by the ticket, works.

Just checked, that's a bug. [(SHARE-332)](https://jira.catrob.at/browse/SHARE-332)

> should we also "steal" the likes/reactions, views, downloads, comments from scratch projects, and add them to our catrobat version 

I only found a deprecated endpoint for the comments (e.g.: https://scratch.mit.edu/site-api/comments/project/142/). If we want to add the comments we would have to check the request their client makes.
Scratch allows to comment on other comments, share doesn't.
Some scratch projects have a lot of views & downloads, those sections on the homepage would then only show scratch projects, and it would also possibly skew with the recommendation algorithm.
I don't think you can see by whom something is liked in scratch (same for favourites).

> What happens if the scratch user updates his project on scratch? Can we somehow check out the new updated version of the project

It would be possible to check if a project is a scratch project when it gets opened with the /project/ url and then check for updates.
Another possibility would be to add a command that checks all scratch projects for updates. (A problem there would be the rate limit of the Scratch API)

> You also have changed almost all shell scripts. Any reason for this "0 lines, but bytes have changed" changes?

The import-container-libs script had CRLF line separators, so it didn't work with linux hosts. Some other scripts didn't have the executable flag.

PS: at least in the old version of the Scratch API 10 request per second were allowed to be made and returned at max 40 results(projects, comments, ...) at once.

dmetzner commented:
> It would be possible to check if a project is a scratch project when it gets opened with the /project/ url and then check for updates.

I would prefer this solution. The project can be updated in the background if necessary + no manual commands or cron jobs are needed.

11possim commented:
Added ScratchController & ScratchManager, moved /scratch_project/{id} to /scratch/project/{id}, allow POST Requests
Added async updating of project when a Scratch project is visited
Implemented requested changes

dmetzner reviewed (approved):
LGTM
